packages/firestore/src/local/document_overlay_cache.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DocumentKey } from '../model/document_key';
+import { Mutation } from '../model/mutation';
+import { Overlay } from '../model/overlay';
+import { ResourcePath } from '../model/path';
+
+import { PersistencePromise } from './persistence_promise';
+import { PersistenceTransaction } from './persistence_transaction';
+
+/**
+ * Provides methods to read and write document overlays.
+ *
+ * An overlay is a saved mutation, that gives a local view of a document when
+ * applied to the remote version of the document.
+ *
+ * Each overlay stores the largest batch ID that is included in the overlay,
+ * which allows us to remove the overlay once all batches leading up to it have
+ * been acknowledged.
+ */
+export interface DocumentOverlayCache {
+  /**
+   * Gets the saved overlay mutation for the given document key.
+   * Returns null if there is no overlay for that key.
+   */
+  getOverlay(
+    transaction: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<Overlay | null>;
+
+  /**
+   * Saves the given document mutation map to persistence as overlays.
+   * All overlays will have their largest batch id set to `largestBatchId`.
+   */
+  saveOverlays(
+    transaction: PersistenceTransaction,
+    largestBatchId: number,
+    overlays: Map<DocumentKey, Mutation>
+  ): PersistencePromise<void>;
+
+  /** Removes the overlay whose largest-batch-id equals the given Id. */
+  removeOverlaysForBatchId(
+    transaction: PersistenceTransaction,
+    batchId: number
+  ): PersistencePromise<void>;
+
+  /**
+   * Returns all saved overlays for the given collection.
+   *
+   * @param transaction The persistence transaction to use for this operation.
```

Comment:
```suggestion
   * @param transaction - The persistence transaction to use for this operation.
```

This applies everywhere. You should be getting warnings during the build for not including the dashes. I hate it too.